kpatch module integration test

This tests several things related to the patching of modules:

- 'kpatch_string' tests the referencing of a symbol which is outside the
  .o, but inside the patch module.

- alternatives patching (.altinstructions)

- paravirt patching (.parainstructions)

- jump labels (5.8+ kernels only) -- including dynamic printk

Signed-off-by: Josh Poimboeuf <jpoimboe@redhat.com>

diff --git a/fs/xfs/xfs_stats.c b/fs/xfs/xfs_stats.c
index 90a77cd3ebad..f65df90ae8e1 100644
--- a/fs/xfs/xfs_stats.c
+++ b/fs/xfs/xfs_stats.c
@@ -16,6 +16,8 @@ static int counter_val(struct xfsstats __percpu *stats, int idx)
 	return val;
 }
 
+extern char *kpatch_string(void);
+
 int xfs_stats_format(struct xfsstats __percpu *stats, char *buf)
 {
 	int		i, j;
@@ -85,6 +87,34 @@ int xfs_stats_format(struct xfsstats __percpu *stats, char *buf)
 		0);
 #endif
 
+	/* Reference a symbol outside the .o yet inside the patch module: */
+	len += scnprintf(buf + len, PATH_MAX-len, "%s\n", kpatch_string());
+
+#ifdef CONFIG_X86_64
+	/* Test alternatives patching: */
+	alternative("ud2", "nop", X86_FEATURE_ALWAYS);
+	alternative("nop", "ud2", X86_FEATURE_IA64);
+
+	/* Test paravirt patching: */
+	slow_down_io();   /* paravirt call */
+#endif
+
+	/* Test pr_debug: */
+	pr_debug("kpatch: pr_debug() test\n");
+
+{
+	/* Test static branches: */
+	static DEFINE_STATIC_KEY_TRUE(kpatch_key);
+
+       if (static_branch_unlikely(&memcg_kmem_enabled_key))
+		printk("kpatch: memcg_kmem_enabled_key\n");
+
+	BUG_ON(!static_branch_likely(&kpatch_key));
+	static_branch_disable(&kpatch_key);
+	BUG_ON(static_branch_likely(&kpatch_key));
+	static_branch_enable(&kpatch_key);
+}
+
 	return len;
 }
 
diff --git a/net/netlink/af_netlink.c b/net/netlink/af_netlink.c
index c64277659753..05b753c2ec84 100644
--- a/net/netlink/af_netlink.c
+++ b/net/netlink/af_netlink.c
@@ -2980,4 +2980,9 @@ static int __init netlink_proto_init(void)
 	panic("netlink_init: Cannot allocate nl_table\n");
 }
 
+char *kpatch_string(void)
+{
+	return "kpatch";
+}
+
 core_initcall(netlink_proto_init);
